Documentation: API reference for portable stories

yannbf opened the pull request:
Closes #25928, #26090, #25879

<!-- If your PR is related to an issue, provide the number(s) above; if it resolves multiple issues, be sure to break them up (e.g. "closes #1000, closes #1001"). -->

<!--

Thank you for contributing to Storybook! Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/contribute

-->

## What I did

This PR introduces API references for the concept of portable stories in Jest, Playwright, and Vitest

<img width="960" alt="" src="https://github.com/storybookjs/storybook/assets/486540/20e6fa29-dbb2-44e3-b788-a5861c654cf3">

## Checklist for Contributors

### Testing

#### Manual testing

1. Follow the steps in the [contributing instructions](https://storybook.js.org/docs/react/contribute/new-snippets#preview-your-work) for this branch, `api-ref-portable-stories`

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [x] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [x] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [x] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

kylegach committed:
WIP: Add portable stories API reference

yannbf commented on docs/api/portable-stories.md:
````diff
@@ -0,0 +1,184 @@
+
+```ts
+// setup-tests.ts
+import { setProjectAnnotations } from '@storybook/react';
+import * as frameworkAnnotations from '@storybook/nextjs/preview';
+import * as addonAnnotations from 'my-addon/preview';
+import * as previewAnnotations from './.storybook/preview';
+
+setProjectAnnotations([previewAnnotations, frameworkAnnotations, addonAnnotations]);
````
```suggestion
import { setProjectAnnotations } from '@storybook/react';
import * as addonAnnotations from 'my-addon/preview';
import * as previewAnnotations from './.storybook/preview';

setProjectAnnotations([previewAnnotations, addonAnnotations]);
```

yannbf replied:
Nextjs would have its own setProjectAnnotations so users don't need to access its preview annotations

yannbf commented on docs/api/portable-stories.md:
```diff
@@ -0,0 +1,184 @@
+
+Type: `string`
+
+TK - What is this?
```
During the story composition, the `exportsName` parameter will be used to create a property called `storyName`. This is useful for snapshot testing and the like, because users can refer to the `storyName` property to distinguish their snapshots. Here's how the name is determined internally:

```ts
const storyName =
    exportsName ||
    storyAnnotations.storyName ||
    storyAnnotations.story?.name ||
    storyAnnotations.name ||
    'unknown';
```

The `composeStory` function is used internally by `composeStories`, and `composeStories` has access to the exports name of the named exports and therefore _passes them down automatically_ to `composeStory`. In this case, users don't have to think about this.

However, when **not** using `composeStories`, but rather `composeStory`, the user is passing just the reference of a story. In this case, they might need to pass a custom exportName property if they do something that needs a specific story name or id. So the exportsName parameter is there in case they want to override it to something else. There was a single use case in a Github issue where someone from MS wanted this badly, so it was implemented: https://github.com/storybookjs/storybook/pull/22471

In general I think people don't even need to know this exists!

kylegach replied:
Cases like this are tricky, where the nuance and details require a lengthy explanation as you've provided (thank you!) while that same lengthy explanation would give it far too much prominence in our docs. Ultimately, I opted to keep it as brief as possible. Do you think this is comprehensive enough?

> You probably don't need this. Because `composeStory` accepts a single story, it does not have access to the name of that story's export in the file (like `composeStories` does). If you must ensure unique story names in your tests and you cannot use `composeStories`, you can pass the name of the story's export here.

yannbf replied:
Yup that sums it up pretty nicely!

yannbf commented on docs/api/portable-stories.md:
```diff
@@ -0,0 +1,184 @@
+import { render, screen } from '@testing-library/react';
+import { composeStory } from '@storybook/react';
+import { vi, test, expect } from 'vitest';
+import meta, { Primary as PrimaryStory } from './Button.stories';
```
Do you think it's better to show users they should rename the imports in comparison to assigning a different name to their constant e.g.
```ts
import meta, { Primary } from './Button.stories';
const PrimaryStory = composeStory(Primary, meta);
```

kylegach replied:
Good point. I like the more straightforward approach you're suggesting here.

yannbf commented on docs/api/portable-stories.md:
```diff
@@ -0,0 +1,184 @@
+
+## setProjectAnnotations
+
+This API should be called once and before the tests run. This will make sure that whenever `composeStories` or `composeStory` are used, the project annotations are taken into account as well.
```
Would it make sense to mention that this is normally used in setup files in testing environments? Or do you think people will just infer from the `// setup-tests.ts` comment?

yannbf commented on docs/api/portable-stories.md:
```diff
@@ -0,0 +1,184 @@
+
+Type: `ProjectAnnotation | ProjectAnnotation[]`
+
+TK - Describe annotations (maybe have a dedicated section at the bottom that this references?)
```
I agree, and in fact we don't mention the term `annotation` properly throughout the docs, so users might not understand this straight away. The only documentation about project annotations I think we could refer to is this:
https://storybook.js.org/docs/configure#configure-story-rendering

yannbf replied:
But I think you could mention that projectAnnotations relate to the `.storybook/preview.ts` file, so users can define overrides that would match the file signature

yannbf commented on docs/api/portable-stories.md:
```diff
@@ -0,0 +1,184 @@
+
+Type: CSF file exports
+
+Specifies which component's stories you want to compose. Pass the **full set of exports** from the CSF file (not the default export!).
```
will people understand "full set of exports"? would it be better to add an example snippet like so?

```ts
import * as stories from './Button.stories';
```

yannbf commented on docs/api/portable-stories.md:
```diff
@@ -0,0 +1,184 @@
+
+test('renders primary button with overriden props', () => {
+  // You can override props and they will get merged with values from the Story's args
+  render(<Primary>Hello world</Primary>);
```
For Vue, this is the way:
```ts
render(Primary({ label: 'Hello world' }))
```

kylegach committed:
Fix `setProjectAnnotations` snippet

Co-authored-by: Yann Braga <yannbf@gmail.com>

kylegach committed:
WIP: More API referencin'

yannbf commented on docs/api/portable-stories.md:
```diff
@@ -0,0 +1,430 @@
+
+A set of project [annotations](#story-annotations) (those defined in `.storybook/preview.js|ts`) or an array of sets of project annotations, which will be applied to all composed stories.
+
+## Story annotations
```
Maybe this should be called annotations? Becuase story annotations could be confused with story-level annotations 🤔 

I feel like there should be a specific docs page about annotations, which mentions about story level, component level (meta), and project level (preview.js). This is an important knowledge in Storybook, not necessarily related to portable stories

kylegach replied:
I changed the heading to just "Annotations".

I agree with you on the dedicated docs page, but that would require a bunch of work to all related docs pages, which is well out of scope of this effort.

yannbf commented on docs/api/portable-stories.md:
```diff
@@ -0,0 +1,430 @@
+});
+
+test('renders in Spanish', async () => {
+  const PrimaryStory = composeStory(Primary, meta, { globals: { locale: 'es' } });
```
little reminder to myself to check this, I am not sure if it's globals or globalTypes

yannbf commented on docs/api/portable-stories.md:
```diff
@@ -0,0 +1,430 @@
+---
+title: 'Portable stories'
```
Somehow we need to show that portable stories are only supported in React and Vue3 (for now). Not sure how to do that in the docs 🤔

yannbf commented on docs/api/story-pipeline.png:
Hey @cdedreuille are you able to assist us with creating a good visual for this? I can explain to you in detail

cdedreuille replied:
Yes, happy to help. Let's chat on Slack.

kylegach replied:
@cdedreuille — Heads up that the rough draft image has been revised slightly: https://github.com/storybookjs/storybook/pull/26081#discussion_r1499462474

kylegach committed:
Fix typo

Co-authored-by: Yann Braga <yannbf@gmail.com>

kylegach committed:
Add conditional content and snippets

yannbf commented on docs/api/doc-block-useof.md:
```diff
@@ -56,18 +56,20 @@ import * as ButtonStories from './Button.stories';
 ## useOf
 
-## Signature
+### Type
```
reminder to extract this to a separate, small PR to `next`

yannbf commented on docs/api/portable-stories.md:
```diff
@@ -0,0 +1,376 @@
+// Replace <your-renderer> with your renderer, e.g. react, vue3
+import { setProjectAnnotations } from '@storybook/<your-renderer>';
+import * as addonAnnotations from 'my-addon/preview';
+import * as previewAnnotations from './.storybook/preview';
```
Actually let me try to do the following:
- support always using `* as` by adding a `.default` check inside of setProjectAnnotations
- support Typescript types inference as well

yannbf commented on docs/api/story-pipeline.png:
Here's the revisited image
<img width="1476" alt="image" src="https://github.com/storybookjs/storybook/assets/1671563/f8f793a8-b9e9-4267-a666-44ad306b25be">

yannbf commented on docs/snippets/react/portable-stories-cypress-ct-async.ts.mdx:
```diff
@@ -0,0 +1,26 @@
+    cy.then(async () => {
+      await CSF3Primary.play({
+        // data-cy-root is Cypress "App". The top-most wrapping element
+        canvasElement: document.querySelector('[data-cy-root]'),
```
Reminder to myself to experiment with adding internal canvasElement via a decorator

yannbf commented on docs/api/portable-stories.md:
```diff
@@ -0,0 +1,376 @@
+
+### Playwright CT
+
+Portable stories work in [Playwright CT](https://playwright.dev/docs/test-components), although with a few modifications, as well as different mounting mechanism.
```
Maybe we use an image like so
<img width="1251" alt="image" src="https://github.com/storybookjs/storybook/assets/1671563/ca7bd677-0816-46c1-9cac-70bcbfeeb99d">

kylegach replied:
@cdedreuille — Here's another one we'll need. We accidentally flipped the order the first two boxes contained within `playwright test`. It should match the order of the other image.

yannbf commented on docs/api/portable-stories.md:
```diff
@@ -0,0 +1,376 @@
+
+<If renderer={SUPPORTED_RENDERERS}>
+
+Portable stories are Storybook [stories](../writing-stories/index.md) which can be used in external environments such as in a test (with Vitest, for instance). You can use the [`composeStories`](#composestories) and [`composeStory`](#composestory) functions to compose stories and their [annotations](#annotations), making them portable.
```
Seems like it would be better to have separate pages, e.g.
- portable stories in Unit testing
- portable stories in Playwright CT
- portable stories in Cypress CT

kylegach committed:
Split into three separate pages

- Organizational and prose improvements
- Remove Cypress, for now

kylegach commented on docs/api/portable-stories-playwright.md:
````diff
@@ -0,0 +1,183 @@
+
+```ts
+createTest(
+  baseTest: TK
+) => TK
````
@yannbf — Not sure about the types here.

yannbf replied:
Great question, these are types from the Playwright's test function, and the return type is also the type of a Playwright's test function, however with a different `mount` argument to that function. @JReinhold do you have any idea on how to present this here (and the TK's right below this comment) in a way that helps users?

JReinhold replied:
So the "correct" type of this would be

```ts
import type { TestType } from '@playwright/experimental-ct-core';

createTest(
  baseTest: TestType
) => TestType
```

But that seems a little verbose for these docs maybe?
I suggest we reference the [Fixture documentation from Playwright](https://playwright.dev/docs/test-fixtures#creating-a-fixture), because that is actually what we do here, we override the base Fixture.

kylegach commented on docs/snippets/react/portable-stories-jest-compose-story.ts.mdx:
```diff
@@ -0,0 +1,19 @@
+  // Returns a story which already contains all annotations from story, meta and global levels
+  const PrimaryStory = composeStory(Primary, meta);
+
+  const onClickSpy = jest.fn();
```
@yannbf — Did we want to use mocked args here (and all the other similar snippets) instead?

yannbf replied:
I guess not, it will probably make things more confusing

kylegach commented on docs/toc.js:
```diff
@@ -691,6 +691,29 @@ module.exports = {
+          type: 'menu',
+          children: [
+            {
+              pathSegment: 'portable-stories-jest',
```
@yannbf — Maybe we don't need `portable-stories` in the URL? `storybook.js.org/docs/jest` vs. `storybook.js.org/docs/portable-stories-jest`. Wdyt?

yannbf replied:
That's a good question, I guess as long as SEO is good for these cases:
- Storybook tests jest
- Storybook portable stories
- Storybook unit tests

it should be good, but I'm no expert.

@winkerVSbecks could you chime in? For context, we are writing docs for the `composeStories` functions which we are trying to surface as "portable stories". There will be multiple pages, one for each use case (Jest, Vitest, Playwright CT). Thanks <3

winkerVSbecks replied:
Will this become the primary integration with those tools going forward? If so, then I would go with `storybook.js.org/docs/jest`, etc. If this is one of multiple ways you can use Storybook with Jest/Vitest/etc. then I would use `portable-stories-jest`, etc.

kylegach replied:
I forgot about the `api` directory in the route, so the best we could do is `storybook.js.org/docs/api/jest`. I think that 1) is not that much better than `storybook.js.org/docs/api/portable-stories-jest`, and 2) warrants a proper `storybook.js.org/jest` page, if we deem it important (in a future effort).

kylegach commented on docs/api/portable-stories-jest.md:
```diff
@@ -0,0 +1,288 @@
+
+<Callout variant="warning">
+
+Sometimes a story can require an addon's [decorator](../writing-stories/decorators.md) to render properly. For example, an addon can apply a decorator that wraps your story in the necessary router context. In this case, you must include that addon's `preview` export in the project annotations set.
```
@yannbf — How's this? It's the same text for all three pages.

yannbf replied:
The tricky part is that this only applies for the following scenario:
- Addons which provide decorators or loaders, but only internally, and those decorators are needed for your story to render properly.

For instance, addon-themes, msw addon, etc, all have decorators/loaders, however not in their preview preset, but exported instead (so the users have to declare them manually). In that case, these annotations will be applied correctly as they are part of the user's .storybook/preview file

kylegach replied:
Revised:

<img width="683" alt="" src="https://github.com/storybookjs/storybook/assets/486540/3ad0bdcf-5d09-4f51-8283-7b08d4181b85">

yannbf replied:
That's great!

yannbf commented on docs/snippets/vue/portable-stories-vitest-override-globals.ts.mdx:
```diff
@@ -0,0 +1,24 @@
+  const PrimaryStory = composeStory(
+    Primary,
+    meta,
+    { globalTypes: { locale: 'en' } } // 👈 Project annotations to override the locale
```
Turns out there was a bug in portable stories which didn't allow users to use `globals` and only use `globalTypes`, but that doesn't really make sense. Now that is fixed, so we need to update to:

```ts
{ globals: { locale: 'en' } }
```

This matches the API of which users are familiar with in a story render function:
```ts
export const MyStory = (args, { globals }) => {}
```

instead of the API for defining toolbars in .storybook/preview (which will also work, but does not feel right):
```
export default {
  globalTypes: {
    locale: {  defaultValue: 'en' }
  }
}
```

yannbf commented on docs/snippets/react/portable-stories-jest-compose-story.ts.mdx:
````diff
@@ -0,0 +1,19 @@
+```tsx
+// Button.test.tsx
+import { jest, test, expect } from 'jest';
````
I think the imports are from `'@jest/globals'`

yannbf commented on docs/api/portable-stories-playwright.md:
```diff
@@ -0,0 +1,183 @@
+
+To preview your stories, Storybook runs a story pipeline, which includes applying project annotations, loading data, rendering the story, and playing interactions. This is a simplified version of the pipeline:
+
+![TK - Proper alt text once proper image is in place](story-pipeline-playwright-ct.png)
```
Reminder comment so we don't forget to change the alt text

kylegach committed:
Address comments

- Improve `addonAnnotations` callout
- Fix Playwright override globals snippet
- Fix jest imports in snippets
- Replace `globalTypes` with `globals` in snippets

kylegach committed:
Add callouts for nextjs

kylegach committed:
Simplify play function calls

kylegach committed:
Update placeholder story pipeline diagrams

JReinhold commented on docs/api/portable-stories-playwright.md:
```diff
@@ -0,0 +1,208 @@
+
+(⚠️ **Experimental**)
+
+Instead of using Playwright's own `test` function, you can use Storybook's special `createTest` function which will extend Playwright's test functionality to add a custom `mount` mechanism which will load, render, and play the story. This function is experimental and is subject to changes.
```
```suggestion
Instead of using Playwright's own `test` function, you can use Storybook's special `createTest` function to extend Playwright's base fixture and override the `mount` function so it loads, renders, and plays the story. This function is experimental and is subject to changes.

Read more on extending base fixtures in [the Playwright documentation](https://playwright.dev/docs/test-fixtures#creating-a-fixture).
```

kylegach replied:
Thanks! I tweaked the wording slightly, so didn't accept the suggestion directly.

kylegach committed:
Add details about vue alias and Playwright types

kylegach committed:
Format snippets

JReinhold commented on docs/api/portable-stories-jest.md:
```diff
@@ -0,0 +1,317 @@
+
+👉 For this, you use the [`composeStories`](#composestories) or [`composeStory`](#composestory) API. The composed Story will return a `play` method to be called **after** it has rendered.
+
+The play function needs a `canvasElement`, which should be passed by you. A `canvasElement` is the HTML element which wraps your component. Each testing utility provides different ways to retrieve such element, but here's how to do it with Testing Library:
```
@yannbf we need to update this now that `canvasElement` is optional/maybe never needed(?)

yannbf replied:
Sure thing!

JReinhold commented on docs/snippets/vue/portable-stories-jest-with-play-function.ts.mdx:
```diff
@@ -0,0 +1,18 @@
+  const PrimaryStory = composeStory(Primary, meta);
+
+  // First, render the story
+  render(PrimaryStory());
```
If https://github.com/storybookjs/storybook/pull/26198 lands as is, the API has changed slightly

```suggestion
  render(PrimaryStory);
```

JReinhold commented on docs/snippets/vue/portable-stories-jest-compose-stories.ts.mdx:
```diff
@@ -0,0 +1,26 @@
+
+test('renders primary button with overriden props', () => {
+  // You can override props and they will get merged with values from the story's args
+  render(Primary({ label: 'Hello world' }));
```
If https://github.com/storybookjs/storybook/pull/26198 lands as-is, the API for rendering Vue components is changed slightly:

```suggestion
  render(Primary, { props: { label: 'Hello world' } });
```

See [`03ca102` (#26198)](https://github.com/storybookjs/storybook/pull/26198/commits/03ca10229e0b786215aaded00ba7577e6265aa9e)

JReinhold reviewed (commented):
Looks good, we need  few minor adjustments to the Vue API though.

Would it make sense to collapse the Jest and Vitest pages into one, to make it easier to maintain? It feels like they are _very_ similar to each other with only minor differences, maybe even just the snippets are different? Just a suggestion, I don't have the full overview of the docs' information architecture.

yannbf committed:
update snippets

yannbf committed:
minor adjustments
